File: operators/tcn_artekmed/tcn_shm_io/ShmSubscriberOp.py
# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

from .shm_serde import shm_transport_enum

log = logging.getLogger("TcnShmReaderOp")

from typing import Any
import queue
from concurrent.futures import Future, ThreadPoolExecutor

import holoscan as hs
import numpy as np
import cupy as cp

from holoscan.conditions import AsynchronousCondition, AsynchronousEventState
from holoscan.core import Operator, OperatorSpec
from holoscan.operators import HolovizOp
from holoscan.operators import holoviz

from holohub.tcn_depthimage_backprojection import TcnDepthImageBackprojectionOp
from holohub.tcn_depthimage_backprojection._tcn_depthimage_backprojection import CameraModel, DistortionType, RigidTransform, CameraParameters, make_rigid_transform


class ShmSubscriberOp(Operator):
    """Simple zenoh subscriber.

    On each tick, it transmits a received message to the "out" port.

    **==Named Outputs==**

        out : bytes
        the received payload.
    """

    def __init__(
            self,
            fragment: Any,
            subscriber: Any,
            stream_name: str,
            channel_config: Any,
            cycle_time_ms: int,
            *args,
            **kwargs,
    ):
        self.subscriber = subscriber
        self.channel_config = channel_config
        self.stream_name = stream_name
        self.cycle_time_ms = cycle_time_ms
        self.pool = None

        self.executor_ = ThreadPoolExecutor(max_workers=1)
        self.future_ = None  # will be set during start()
        self.async_cond_ = AsynchronousCondition(fragment, name="async_cond")
        self.buffer = queue.Queue()

        # Need to call the base class constructor last
        super().__init__(fragment, self.async_cond_, *args, **kwargs)

    def on_receive(self, user_header: Any, message: Any):
        """Function to be supplied as callback

        When the condition's event_state is EVENT_WAITING, set to EVENT_DONE. This function will
        only exit once the condition is set to EVENT_NEVER.
        """
        if self.async_cond_.event_state == AsynchronousEventState.EVENT_NEVER:
            return False

        # log.debug(f"on_receive frame with timestamp {user_header.timestamp}")
        color_data = {}
        depth_data = {}
        frame_timestamp = user_header.timestamp
        for port in message.ports:
            if port.data.portType == shm_transport_enum.CameraPortType.colorimage:
                # log.debug(f"added port data: {port.name}")
                md = port.data.metadata
                mv = memoryview(port.data.data)
                color_data[port.name] = cp.asarray(np.frombuffer(mv, dtype=np.uint8).reshape(
                    md.header.dimY, md.header.dimX, int(md.header.bitsPerElement/8)
                ))
            elif port.data.portType == shm_transport_enum.CameraPortType.depthimage:
                # log.debug(f"added port data: {port.name}")
                md = port.data.metadata
                mv = memoryview(port.data.data)
                depth_data[port.name] = cp.asarray(np.frombuffer(mv, dtype=np.uint16).reshape(
                    md.header.dimY, md.header.dimX, 1
                ))

        if color_data or depth_data:
            # how does ts relate to fragment.scheduler().clock.timestamp()?
            # log.debug(f"put data for {frame_timestamp} into queue")
            self.buffer.put((frame_timestamp, (color_data, depth_data)))

            if self.async_cond_.event_state == AsynchronousEventState.EVENT_WAITING:
                self.async_cond_.event_state = AsynchronousEventState.EVENT_DONE
            return True

        return False

    def receiver_mainloop(self):
        while self.subscriber is not None:
            if not self.subscriber.receive_frame(self.on_receive, self.cycle_time_ms):
                log.warning("could not receive frame.")

    def setup(self, spec: OperatorSpec):
        spec.output("color_outputs")
        spec.output("color_output_specs").condition(hs.core.ConditionType.NONE)
        spec.output("depth_outputs")
        spec.output("depth_output_specs").condition(hs.core.ConditionType.NONE)

    def start(self):
        self.subscriber.subscribe(self.stream_name)

        self.future_ = self.executor_.submit(self.receiver_mainloop)
        assert isinstance(self.future_, Future)

    def compute(self, op_input, op_output, context):
        scheduler = self.fragment.scheduler()
        clock = scheduler.clock
        ts = clock.timestamp()

        frame_ts, (color_data, depth_data) = self.buffer.get()
        log.debug(f"got data for {frame_ts} from queue")

        color_message = {k:hs.as_tensor(v) for k,v in color_data.items()}
        depth_message = {k:hs.as_tensor(v) for k,v in depth_data.items()}

        self.async_cond_.event_state = AsynchronousEventState.EVENT_WAITING
        op_output.emit(color_message, "color_outputs", acq_timestamp=ts)
        op_output.emit(depth_message, "depth_outputs", acq_timestamp=ts)

        num_videos = len(color_message.keys())
        # assume they are the same
        #num_depth_videos = len(depth_message.keys())


        # Determine grid size (e.g., 2 for 2x2, 3 for 3x3)
        grid_size = int(np.ceil(np.sqrt(num_videos))) if num_videos > 0 else 1
        tile_size = 1.0 / grid_size

        color_output_specs = []
        for i, port_name in enumerate(color_message.keys()):
            # Compute row and column index
            row = i // grid_size
            col = i % grid_size

            # Compute normalized offsets (0.0 to 1.0)
            # Note: Holoviz usually uses (x, y) for offsets
            offset_x = col * tile_size
            offset_y = row * tile_size

            spec = HolovizOp.InputSpec(port_name, HolovizOp.InputType.COLOR)
            views = []
            view = HolovizOp.InputSpec.View()
            view.offset_x = offset_x
            view.offset_y = offset_y
            view.width = tile_size
            view.height = tile_size
            views.append(view)
            spec.views = views
            # hardcoded ..
            spec.image_format = holoviz._holoviz_str_to_image_format["b8g8r8a8_unorm"]
            color_output_specs.append(spec)

        op_output.emit(color_output_specs, "color_output_specs", acq_timestamp=ts)

        depth_output_specs = []
        for i, port_name in enumerate(depth_message.keys()):
            # Compute row and column index
            row = i // grid_size
            col = i % grid_size

            # Compute normalized offsets (0.0 to 1.0)
            # Note: Holoviz usually uses (x, y) for offsets
            offset_x = col * tile_size
            offset_y = row * tile_size

            # still using color as uint16 is not supported as depth-format
            spec = HolovizOp.InputSpec(port_name, HolovizOp.InputType.COLOR)
            views = []
            view = HolovizOp.InputSpec.View()
            view.offset_x = offset_x
            view.offset_y = offset_y
            view.width = tile_size
            view.height = tile_size
            views.append(view)
            spec.views = views
            depth_output_specs.append(spec)

        op_output.emit(depth_output_specs, "depth_output_specs", acq_timestamp=ts)


    def stop(self):
        self.async_cond_.event_state = AsynchronousEventState.EVENT_NEVER
        if self.subscriber is not None:
            self.subscriber = None
        self.future_.result()

File: operators/tcn_artekmed/tcn_shm_subscriber/shm_subscriber_op.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "shm_subscriber_op.hpp"

#include <holoscan/utils/cuda_macros.hpp>

#include "gxf/std/timestamp.hpp"  // nvidia::gxf::Timestamp -- acquisition time for frame grouping

#include "../common/utils.h"

namespace tcn::ops {

void TcnShmSubscriberOp::setup(holoscan::OperatorSpec& spec) {
    spec.output<holoscan::gxf::Entity>("color_outputs");
    spec.output<holoscan::gxf::Entity>("depth_outputs");

    spec.param(stream_name_, "stream_name",
               "Stream Name",
               "SHM stream name to subscribe to",
               std::string{});
    spec.param(cycle_time_ms_, "cycle_time_ms",
               "Cycle Time (ms)",
               "Wait time between frame polls in milliseconds",
               int32_t{1});
    spec.param(allocator_, "allocator",
               "Allocator",
               "Memory allocator for output tensors");
    spec.param(async_condition_, "async_condition",
               "Async Condition",
               "AsynchronousCondition for scheduling");
}

void TcnShmSubscriberOp::initialize() {
    // Register the AsynchronousCondition as a scheduling condition
    if (async_condition_.has_value()) {
        add_arg(async_condition_.get());
    }
    Operator::initialize();
}

void TcnShmSubscriberOp::start() {
    if (!receiver_) {
        HOLOSCAN_LOG_ERROR("ShmSubscriberOp: no receiver set. "
                           "Set receiver_ before starting the operator.");
        return;
    }

    // Subscribe to the frame data stream
    auto name = stream_name_.get();
    if (!receiver_->subscribe(name)) {
        HOLOSCAN_LOG_ERROR("Failed to subscribe to stream: {}", name);
        return;
    }

    // Create a dedicated CUDA stream for SHM→GPU async copies
    HOLOSCAN_CUDA_CALL(cudaStreamCreateWithFlags(&copy_stream_, cudaStreamNonBlocking));

    should_stop_.store(false);

    // Set async condition to waiting state
    if (async_condition_.get()) {
        async_condition_.get()->event_state(holoscan::AsynchronousEventState::EVENT_WAITING);
    }

    // Launch background receiver thread
    receiver_thread_ = std::thread(&TcnShmSubscriberOp::receiver_mainloop, this);

    HOLOSCAN_LOG_INFO("ShmSubscriberOp started for stream: {} (zero-copy)", name);
}

void TcnShmSubscriberOp::stop() {
    should_stop_.store(true);

    // Signal the async condition to stop
    if (async_condition_.get()) {
        async_condition_.get()->event_state(holoscan::AsynchronousEventState::EVENT_NEVER);
    }

    // Wait for receiver thread to finish
    if (receiver_thread_.joinable()) {
        receiver_thread_.join();
    }

    // Drain the queue to release any held SHM segments
    {
        std::lock_guard<std::mutex> lock(queue_mutex_);
        while (!frame_queue_.empty()) {
            frame_queue_.pop();
        }
    }

    // Destroy CUDA stream
    if (copy_stream_) {
        cudaStreamSynchronize(copy_stream_);
        cudaStreamDestroy(copy_stream_);
        copy_stream_ = nullptr;
    }

    // Clean up receiver
    if (receiver_) {
        receiver_->teardown();
    }

    HOLOSCAN_LOG_INFO("ShmSubscriberOp stopped");
}

void TcnShmSubscriberOp::receiver_mainloop() {
    last_stats_log_ = std::chrono::steady_clock::now();

    while (!should_stop_.load()) {
        auto frame = receiver_->receive_frame_zero_copy(cycle_time_ms_.get());
        if (!frame.has_value()) {
            continue;
        }

        // Check if we should stop before queuing
        if (async_condition_.get() &&
            async_condition_.get()->event_state() == holoscan::AsynchronousEventState::EVENT_NEVER) {
            break;
        }

        frames_received_.fetch_add(1, std::memory_order_relaxed);

        {
            std::lock_guard<std::mutex> lock(queue_mutex_);
            // Back-pressure: if queue is full, drop oldest frame to keep
            // at most kMaxQueuedFrames in-flight (releases the SHM segment).
            while (frame_queue_.size() >= kMaxQueuedFrames) {
                frame_queue_.pop();
                frames_skipped_.fetch_add(1, std::memory_order_relaxed);
            }
            frame_queue_.push(std::move(*frame));
        }

        // Periodic stats log (every 5 seconds)
        auto now = std::chrono::steady_clock::now();
        if (now - last_stats_log_ >= std::chrono::seconds(5)) {
            auto skipped = frames_skipped_.exchange(0, std::memory_order_relaxed);
            auto received = frames_received_.exchange(0, std::memory_order_relaxed);
            if (skipped > 0) {
                HOLOSCAN_LOG_WARN("SHM receiver: {}/{} frames skipped in last 5s",
                                  skipped, received);
            } else {
                HOLOSCAN_LOG_INFO("SHM receiver: {}/{} frames processed in last 5s",
                                  received, received);
            }
            last_stats_log_ = now;
        }

        // Notify the Holoscan scheduler
        if (async_condition_.get() &&
            async_condition_.get()->event_state() == holoscan::AsynchronousEventState::EVENT_WAITING) {
            async_condition_.get()->event_state(holoscan::AsynchronousEventState::EVENT_DONE);
        }
    }
}

void TcnShmSubscriberOp::compute(
    holoscan::InputContext& op_input,
    holoscan::OutputContext& op_output,
    holoscan::ExecutionContext& context) {

    // Pop a zero-copy frame from the queue
    tcn::shm::ShmZeroCopyFrame frame;
    {
        std::lock_guard<std::mutex> lock(queue_mutex_);
        if (frame_queue_.empty()) {
            HOLOSCAN_LOG_WARN("compute() called with empty frame queue");
            if (async_condition_.get()) {
                async_condition_.get()->event_state(holoscan::AsynchronousEventState::EVENT_WAITING);
            }
            return;
        }
        frame = std::move(frame_queue_.front());
        frame_queue_.pop();
    }

    HOLOSCAN_LOG_DEBUG("Processing frame ts={} (zero-copy, {} ports)",
                       frame.timestamp, frame.ports.size());

    // Capture the acquisition timestamp BEFORE `frame` is cleared further down (`frame = {}` runs
    // before the emits, to release the SHM segment as early as possible), otherwise this reads 0.
    //
    // Units are nanoseconds, but the EPOCH is the publisher's std::chrono::steady_clock
    // (tcn_shm_zenoh_sender: user_header.timestamp = now_ns), NOT the GXF global clock and not wall
    // time. So these values are only meaningful RELATIVE TO EACH OTHER: they are safe to compare
    // and match across streams from the same publisher, and must never be compared against a
    // locally computed "now" or against a second publisher's timestamps.
    //
    // One timestamp per composite buffer, shared by every camera port in it -- the cameras in a
    // segment are synchronised at capture and ShmPortView carries no per-port time. So this is a
    // capture-GROUP identity, which is the granularity tcn_temporal_sync matches on.
    const int64_t acq_timestamp_ns = static_cast<int64_t>(frame.timestamp);

    // Get allocator handle for tensor allocation
    auto allocator_handle = nvidia::gxf::Handle<nvidia::gxf::Allocator>::Create(
        context.context(), allocator_->gxf_cid());
    if (!allocator_handle) {
        HOLOSCAN_LOG_ERROR("Failed to get allocator handle");
        return;
    }

    // Create output entities
    auto color_entity = nvidia::gxf::Entity::New(context.context());
    if (!color_entity) {
        HOLOSCAN_LOG_ERROR("Failed to create color entity");
        return;
    }
    auto depth_entity = nvidia::gxf::Entity::New(context.context());
    if (!depth_entity) {
        HOLOSCAN_LOG_ERROR("Failed to create depth entity");
        return;
    }

    // For each port, allocate a GPU tensor and kick off an async copy
    // directly from the SHM pointer to GPU device memory.
    for (const auto& pv : frame.ports) {
        if (pv.is_color) {
            nvidia::gxf::Handle<nvidia::gxf::Tensor> tensor;
            if (!tcn::allocate_named_tensor<uint8_t>(
                    allocator_handle.value(), copy_stream_, color_entity.value(),
                    nvidia::gxf::Shape{{pv.height, pv.width, pv.channels}},
                    nvidia::gxf::MemoryStorageType::kDevice,
                    pv.name, tensor)) {
                HOLOSCAN_LOG_ERROR("Failed to allocate color tensor for {}", pv.name);
                continue;
            }

            auto maybe_data = tensor->data<uint8_t>();
            if (maybe_data) {
                HOLOSCAN_CUDA_CALL(cudaMemcpyAsync(
                    maybe_data.value(), pv.data_ptr, pv.data_size,
                    cudaMemcpyHostToDevice, copy_stream_));
            }
        } else {
            nvidia::gxf::Handle<nvidia::gxf::Tensor> tensor;
            if (!tcn::allocate_named_tensor<uint16_t>(
                    allocator_handle.value(), copy_stream_, depth_entity.value(),
                    nvidia::gxf::Shape{{pv.height, pv.width, pv.channels}},
                    nvidia::gxf::MemoryStorageType::kDevice,
                    pv.name, tensor)) {
                HOLOSCAN_LOG_ERROR("Failed to allocate depth tensor for {}", pv.name);
                continue;
            }

            auto maybe_data = tensor->data<uint16_t>();
            if (maybe_data) {
                HOLOSCAN_CUDA_CALL(cudaMemcpyAsync(
                    maybe_data.value(), pv.data_ptr, pv.data_size,
                    cudaMemcpyHostToDevice, copy_stream_));
            }
        }
    }

    // Wait for all async copies to complete before releasing the SHM segment.
    // This is the critical safety barrier: SHM data_ptrs are valid only while
    // frame.shm_handle is alive, and the CUDA DMA engine reads from those
    // addresses asynchronously.
    HOLOSCAN_CUDA_CALL(cudaStreamSynchronize(copy_stream_));

    // Release SHM segment — frame goes out of scope at function end, but
    // we explicitly clear it here to make the release point obvious and to
    // ensure it happens before the emit (defense in depth).
    frame = {};

    // Stamp both entities with the acquisition time so downstream operators can group frames that
    // belong together (tcn_temporal_sync). Holoscan already reads this back on the receive side --
    // InputContext::get_acquisition_timestamp() searches a received entity for ANY component of
    // type nvidia::gxf::Timestamp, so the component NAME is not significant; only the type is.
    // Nothing in Holoscan CREATES one, which is why this has to be done here: without it the
    // frame's identity is lost at the source and there is nothing downstream to match on.
    const int64_t pub_timestamp_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
    auto stamp = [&](nvidia::gxf::Entity& e, const char* what) {
        auto ts = e.add<nvidia::gxf::Timestamp>("timestamp");
        if (!ts) {
            // Non-fatal: the frame is still valid data, but anything downstream that groups by
            // acquisition time will not see this one. Warn rather than drop the frame.
            HOLOSCAN_LOG_WARN("Failed to add Timestamp to {} entity (acq={})", what,
                              acq_timestamp_ns);
            return;
        }
        ts.value()->acqtime = acq_timestamp_ns;
        ts.value()->pubtime = pub_timestamp_ns;
    };
    stamp(color_entity.value(), "color");
    stamp(depth_entity.value(), "depth");

    // Emit outputs
    op_output.emit(color_entity.value(), "color_outputs");
    op_output.emit(depth_entity.value(), "depth_outputs");

    // Reset async condition to wait for next frame
    if (async_condition_.get()) {
        async_condition_.get()->event_state(holoscan::AsynchronousEventState::EVENT_WAITING);
    }
}

}  // namespace tcn::ops
